Use of `atomicAddDouble` crashed the kernel

Below is the output

```
CHIP debug [TID 20420] [1673650037.941151711] : Reading CHIP_PLATFORM from env
CHIP debug [TID 20420] [1673650037.941203491] : Reading CHIP_DEVICE_TYPE from env
CHIP debug [TID 20420] [1673650037.941229203] : Reading CHIP_DEVICE from env
CHIP debug [TID 20420] [1673650037.941233236] : Reading CHIP_BE from env
CHIP debug [TID 20420] [1673650037.941256881] : CHIP_PLATFORM=0
CHIP debug [TID 20420] [1673650037.941260296] : CHIP_DEVICE_TYPE=gpu
CHIP debug [TID 20420] [1673650037.941264000] : CHIP_DEVICE=0
CHIP debug [TID 20420] [1673650037.941267211] : CHIP_BE=default
CHIP debug [TID 20420] [1673650037.941271875] : CHIPDriver Initialize
CHIP debug [TID 20420] [1673650037.941279317] : CHIPBE=default... trying Level0 Backend
CHIP debug [TID 20420] [1673650037.941313740] : CHIPBackend Base Constructor
CHIP debug [TID 20420] [1673650037.958707903] : CHIPQueue() 0x55fe5186d9b0
CHIP debug [TID 20420] [1673650037.958962800] : 0x55fe51781970 CHIPContext::addDevice() 0x55fe5186d2c0
CHIP debug [TID 20420] [1673650037.959048912] : Thread Created with ID : 140532041545472
CHIP debug [TID 20420] [1673650037.959073676] : Thread Created with ID : 140532033152768
CHIP debug [TID 20420] [1673650037.959087827] : Reading CHIP_JIT_FLAGS from env
CHIP debug [TID 20420] [1673650037.959095200] : __hipRegisterFatBinary
CHIP debug [TID 20420] [1673650037.959182415] : Bundle entry ID 0 is: 'host-x86_64-unknown-linux'

CHIP debug [TID 20420] [1673650037.959206439] : Not a SPIR-V triple, ignoring

CHIP debug [TID 20420] [1673650037.959210681] : Bundle entry ID 1 is: 'hip-spirv64----generic'

CHIP debug [TID 20420] [1673650037.959780918] : Register module: 0x55fe5186be00

CHIP debug [TID 20420] [1673650037.959785999] : CHIPBackend->register_module()
CHIP debug [TID 20420] [1673650037.959824046] : RegisterFunction on module 0x55fe5186be00

CHIP debug [TID 20420] [1673650037.959830689] : RegisterFunction on 1 devices
CHIP debug [TID 20420] [1673650037.959833801] : CHIPBackend.registerFunctionAsKernel()
CHIP debug [TID 20420] [1673650037.960799345] : JIT compiler flags: -cl-std=CL2.0 -cl-take-global-address -cl-match-sincospi
CHIP error [TID 20420] [1673650038.048763196] : ZE Build Log:
error: undefined reference to `'
in function: '' called by kernel: 'process_batch_kernel(double, dbm_task_t const*, double const*, double const*, double*)'

error: backend compiler failed build.

error: undefined reference to `'
in function: '' called by kernel: 'process_batch_kernel(double, dbm_task_t const*, double const*, double const*, double*)'

error: backend compiler failed build.

CHIP error [TID 20420] [1673650038.048850190] : hipErrorTbd (ZE_RESULT_ERROR_MODULE_BUILD_FAILURE ) in /home/pvelesko/chip-spv/main/src/backend/Level0/CHIPBackendLevel0.cc:2145:compile

CHIP error [TID 20420] [1673650038.048978032] : hipErrorTbd
CHIP debug [TID 20420] [1673650038.049033917] : __hipRegisterFatBinary
CHIP debug [TID 20420] [1673650038.049044308] : Bundle entry ID 0 is: 'host-x86_64-unknown-linux'

CHIP debug [TID 20420] [1673650038.049048086] : Not a SPIR-V triple, ignoring

CHIP debug [TID 20420] [1673650038.049051574] : Bundle entry ID 1 is: 'hip-spirv64----generic'

CHIP debug [TID 20420] [1673650038.049079080] : Register module: 0x55fe51ad7600

CHIP debug [TID 20420] [1673650038.049096762] : CHIPBackend->register_module()
CHIP debug [TID 20420] [1673650038.049214898] : JIT compiler flags: -cl-std=CL2.0 -cl-take-global-address -cl-match-sincospi
CHIP debug [TID 20420] [1673650038.100844606] : __hipRegisterFatBinary
CHIP debug [TID 20420] [1673650038.100901023] : Bundle entry ID 0 is: 'host-x86_64-unknown-linux'

CHIP debug [TID 20420] [1673650038.100904898] : Not a SPIR-V triple, ignoring

CHIP debug [TID 20420] [1673650038.100908345] : Bundle entry ID 1 is: 'hip-spirv64----generic'

CHIP debug [TID 20420] [1673650038.100939824] : Register module: 0x55fe522541a0

CHIP debug [TID 20420] [1673650038.100943808] : CHIPBackend->register_module()
CHIP debug [TID 20420] [1673650038.101002690] : JIT compiler flags: -cl-std=CL2.0 -cl-take-global-address -cl-match-sincospi
CHIP debug [TID 20420] [1673650038.148943369] : __hipRegisterFatBinary
CHIP debug [TID 20420] [1673650038.149006718] : Bundle entry ID 0 is: 'host-x86_64-unknown-linux'

CHIP debug [TID 20420] [1673650038.149010851] : Not a SPIR-V triple, ignoring

CHIP debug [TID 20420] [1673650038.149014300] : Bundle entry ID 1 is: 'hip-spirv64----generic'

CHIP debug [TID 20420] [1673650038.149031509] : Register module: 0x55fe5239f450

CHIP debug [TID 20420] [1673650038.149035233] : CHIPBackend->register_module()
CHIP debug [TID 20420] [1673650038.149121628] : JIT compiler flags: -cl-std=CL2.0 -cl-take-global-address -cl-match-sincospi
CHIP debug [TID 20420] [1673650038.196785727] : __hipRegisterFatBinary
CHIP debug [TID 20420] [1673650038.196818952] : Bundle entry ID 0 is: 'host-x86_64-unknown-linux'

CHIP debug [TID 20420] [1673650038.196842168] : Not a SPIR-V triple, ignoring

CHIP debug [TID 20420] [1673650038.196845567] : Bundle entry ID 1 is: 'hip-spirv64----generic'

CHIP debug [TID 20420] [1673650038.196875133] : Register module: 0x55fe5350db10

CHIP debug [TID 20420] [1673650038.196878840] : CHIPBackend->register_module()
CHIP debug [TID 20420] [1673650038.196937587] : JIT compiler flags: -cl-std=CL2.0 -cl-take-global-address -cl-match-sincospi
OpenMP-threads: 8  GPUs: 1  Libxsmm: n/a  MPI: n/a

CHIP debug [TID 20420] [1673650038.608409984] : CHIPQueue() 0x55fe51a9db20
CHIP debug [TID 20420] [1673650038.608593089] : 0x55fe5186d2c0 CHIPDevice::addQueue(0x55fe51a9db20)
CHIP debug [TID 20420] [1673650038.608669590] : CHIPQueue 0x55fe51a9db20 added to the queue vector for device 0x55fe5186d2c0
CHIP debug [TID 20420] [1673650038.608734268] : Currently used memory on dev Intel(R) Iris(R) Pro Graphics P580 [0x193a]: 1 M

CHIP debug [TID 20420] [1673650038.608869648] : CHIPAllocationTracker::recordAllocation size: 1920000 HOST 0x0 DEV 0xffffd556aa610000 TYPE 1
CHIP debug [TID 20420] [1673650038.608905840] : CHIPQueue() 0x55fe51ae1d10
CHIP debug [TID 20420] [1673650038.608966655] : 0x55fe5186d2c0 CHIPDevice::addQueue(0x55fe51ae1d10)
CHIP debug [TID 20420] [1673650038.609000620] : CHIPQueue 0x55fe51ae1d10 added to the queue vector for device 0x55fe5186d2c0
CHIP debug [TID 20420] [1673650038.609033189] : Unregister module: 0x55fe5350db10

CHIP debug [TID 20420] [1673650038.609054697] : CHIPBackend->unregister_module()
CHIP debug [TID 20420] [1673650038.609098349] : __hipUnRegisterFatBinary 4


CHIP debug [TID 20420] [1673650038.609125643] : Unregister module: 0x55fe5239f450

CHIP debug [TID 20420] [1673650038.609146188] : CHIPBackend->unregister_module()
CHIP debug [TID 20420] [1673650038.609169914] : __hipUnRegisterFatBinary 3

CHIP debug [TID 20420] [1673650038.609190134] : Unregister module: 0x55fe522541a0

CHIP debug [TID 20420] [1673650038.609210340] : CHIPBackend->unregister_module()
CHIP debug [TID 20420] [1673650038.609232251] : __hipUnRegisterFatBinary 2

CHIP debug [TID 20420] [1673650038.609257169] : Unregister module: 0x55fe51ad7600

CHIP debug [TID 20420] [1673650038.609278522] : CHIPBackend->unregister_module()
CHIP debug [TID 20420] [1673650038.609298982] : __hipUnRegisterFatBinary 1

CHIP debug [TID 20420] [1673650038.609327512] : Unregister module: 0x55fe5186be00

CHIP debug [TID 20420] [1673650038.609353316] : CHIPBackend->unregister_module()
CHIP debug [TID 20420] [1673650038.609375115] : __hipUnRegisterFatBinary 0

CHIP debug [TID 20420] [1673650038.609396491] : Uninitializing CHIP...
CHIP debug [TID 20420] [1673650039.109707118] : Setting LastEvent for 0x55fe5186d9b0 0x0 -> 0x0
CHIP warning [TID 20420] [1673650039.109761853] : Not all user created streams have been destoyed... Queues remaining: 2
CHIP warning [TID 20420] [1673650039.109768875] : Make sure to call hipStreamDestroy() for all queues that have been created via hipStreamCreate()
CHIP warning [TID 20420] [1673650039.109771280] : Removing user-created streams without calling a destructor
CHIP debug [TID 20420] [1673650039.148248635] : CHIPBackend Destructor. Deleting all pointers.
CHIP debug [TID 20420] [1673650039.148351269] : 0x55fe51781970 CHIPContext.removeDevice() 0x55fe5186d2c0
CHIP debug [TID 20420] [1673650039.148370194] : ~CHIPDevice() 0x55fe5186d2c0
CHIP debug [TID 20420] [1673650039.148382915] : Setting LastEvent for 0x55fe5186d9b0 0x0 -> 0x0
CHIP debug [TID 20420] [1673650039.148391376] : Setting LastEvent for 0x55fe5186d9b0 0x0 -> 0x0
CHIP debug [TID 20420] [1673650039.148491189] : ~CHIPContext() 0x55fe51781970
From offload/offload_runtime.h:l115
```

Made a PR (#311 ) for preserving symbols. Here is the improved build log:

```
CHIP error [TID 1997600] [1674038028.713976065] : ZE Build Log: 
error: undefined reference to `_ZL20__longlong_as_doublex'
in function: '__longlong_as_double(long long)' called by kernel: 'repro_kernel(double*, double)'

error: backend compiler failed build.

error: undefined reference to `_ZL20__double_as_longlongd'
in function: '__double_as_longlong(double)' called by kernel: 'repro_kernel(double*, double)'

error: backend compiler failed build.
```
It seems our `__longlong_as_double()` and `__double_as_longlong()` definitions are incomplete.

--- a/bitcode/devicelib.cl
+++ b/bitcode/devicelib.cl
@@ -850,3 +850,7 @@ EXPORT float CL_NAME2(tex2D, f)(hipTextureObject_t textureObject,
     __builtin_astype(textureObject->sampler, sampler_t),
     (float2)(x, y)).x;
 }
+
+// In HIP long long is 64-bit integer. In OpenCL it's 128-bit integer.
+EXPORT long __double_as_longlong(double x) { return as_long(x); }
+EXPORT double __longlong_as_double(long int x) { return as_double(x); }

--- a/cmake/UnitTests.cmake
+++ b/cmake/UnitTests.cmake
@@ -1189,6 +1189,7 @@ list(APPEND DGPU_OPENCL_FAILED_TESTS "Unit_hipEvent") # Failed
 list(APPEND DGPU_OPENCL_FAILED_TESTS "hipStreamSemantics") # Failed
 list(APPEND DGPU_OPENCL_FAILED_TESTS "Unit_hipMemsetFunctional_PartialSet_3D") # Failed
 list(APPEND DGPU_OPENCL_FAILED_TESTS "hipMultiThreadAddCallback") # Subprocess aborted
+list(APPEND DGPU_OPENCL_FAILED_TESTS "Unit_hipMemset3DAsync_ConcurrencyMthread") # Timeout
 
 # # dGPU Level Zero Unit Test Failures
 list(APPEND DGPU_LEVEL0_FAILED_TESTS "hipDynamicShared") # SEGFAULT

--- a/include/hip/devicelib/type_casting_intrinsics.hh
+++ b/include/hip/devicelib/type_casting_intrinsics.hh
@@ -48,7 +48,6 @@ EXPORT unsigned long long int __double2ull_rd(double x);
 EXPORT unsigned long long int __double2ull_rn(double x);
 EXPORT unsigned long long int __double2ull_ru(double x);
 EXPORT unsigned long long int __double2ull_rz(double x);
-EXPORT long long int __double_as_longlong(double x);
 EXPORT int __float2int_rd(float x);
 EXPORT int __float2int_rn(float x);
 EXPORT int __float2int_ru(float);
@@ -82,7 +81,6 @@ EXPORT float __ll2float_rd(long long int x);
 EXPORT float __ll2float_rn(long long int x);
 EXPORT float __ll2float_ru(long long int x);
 EXPORT float __ll2float_rz(long long int x);
-EXPORT double __longlong_as_double(long long int x);
 EXPORT double __uint2double_rn(unsigned int x);
 EXPORT float __uint2float_rd(unsigned int x);
 EXPORT float __uint2float_rn(unsigned int x);
@@ -98,21 +96,9 @@ EXPORT float __ull2float_rn(unsigned long long int x);
 EXPORT float __ull2float_ru(unsigned long long int x);
 EXPORT float __ull2float_rz(unsigned long long int x);
 
-#if defined(__HIP_DEVICE_COMPILE__)
-EXPORT long long int __double_as_longlong(double x) {
-  static_assert(sizeof(long long) == sizeof(double), "");
-  long long Tmp;
-  __builtin_memcpy(&Tmp, &x, sizeof(Tmp));
-  return Tmp;
+extern "C" {
+__device__ long long int __double_as_longlong(double x);
+__device__ double __longlong_as_double(long long int x);
 }
-
-EXPORT double __longlong_as_double(long long int x) {
-  static_assert(sizeof(double) == sizeof(long long), "");
-  double Tmp;
-  __builtin_memcpy(&Tmp, &x, sizeof(Tmp));
-  return Tmp;
-}
-
-#endif
 
 #endif // include guard

--- a/include/hip/devicelib/type_casting_intrinsics.hh
+++ b/include/hip/devicelib/type_casting_intrinsics.hh
@@ -98,4 +98,21 @@ EXPORT float __ull2float_rn(unsigned long long int x);
 EXPORT float __ull2float_ru(unsigned long long int x);
 EXPORT float __ull2float_rz(unsigned long long int x);
 
+#if defined(__HIP_DEVICE_COMPILE__)
+EXPORT long long int __double_as_longlong(double x) {
+  static_assert(sizeof(long long) == sizeof(double), "");
+  long long Tmp;
+  __builtin_memcpy(&Tmp, &x, sizeof(Tmp));
+  return Tmp;
+}
+
+EXPORT double __longlong_as_double(long long int x) {
+  static_assert(sizeof(double) == sizeof(long long), "");
+  double Tmp;
+  __builtin_memcpy(&Tmp, &x, sizeof(Tmp));
+  return Tmp;
+}
+
+#endif
+
 #endif // include guard

--- a/tests/runtime/CMakeLists.txt
+++ b/tests/runtime/CMakeLists.txt
@@ -40,3 +40,4 @@ endfunction()
 
 add_hip_runtime_test(TestLazyModuleInit.cpp)
 add_hip_runtime_test(TestKernelArgs.hip)
+add_hip_runtime_test(RegressionTest302.hip)

--- a/tests/runtime/RegressionTest302.hip
+++ b/tests/runtime/RegressionTest302.hip
@@ -1,0 +1,19 @@
+#include "hip/hip_runtime.h"
+
+// CHIP-SPV/chip-spv#302
+__global__ static void repro_kernel(double *address, double val) {
+  unsigned long long int *address_as_ull = (unsigned long long int *)address;
+  unsigned long long int value_as_ull = *address_as_ull;
+  atomicCAS(address_as_ull, value_as_ull,
+            __double_as_longlong(val + __longlong_as_double(value_as_ull)));
+}
+
+int main() {
+  double *mem_d, mem_h = 1.5;
+  (void)hipMalloc(&mem_d, sizeof(double));
+  (void)hipMemcpy(mem_d, &mem_h, sizeof(double), hipMemcpyHostToDevice);
+  repro_kernel<<<1, 1>>>(mem_d, 1.0);
+  (void)hipMemcpy(&mem_h, mem_d, sizeof(double), hipMemcpyDeviceToHost);
+  (void)hipFree(mem_d);
+  return !(mem_h == 2.5f);
+}

